File: android/app/src/main/kotlin/com/baseapp/android/view/billing/SubscribeView.kt
package com.baseapp.android.view.billing

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.baseapp.android.components.components
import com.baseapp.android.components.store
import com.baseapp.android.view.theme.LocalUiTheme

/**
 * The subscribe flow — the native twin of web/app's SubscribePage.tsx (iOS:
 * SubscribeView.swift): on first composition it starts a subscription
 * checkout for the signed-in user and hands them to the session's checkoutUrl
 * (Stripe's hosted page when deployed, the web app's in-app test checkout in
 * the sandbox) in the system browser, matching how SignInView opens OAuth
 * URLs. Subscription checkout is never anonymous — this view only renders
 * inside the signed-in shell (the web twin routes signed-out visitors to
 * sign-up first).
 *
 * `productKey` picks a catalog plan; null means the default plan (the web
 * twin's `?plan=<key>` query parameter). Present it from a pricing CTA or the
 * Settings Billing card's subscribe affordance.
 */
@Composable
fun SubscribeView(
    productKey: String? = null,
    onDismiss: () -> Unit,
) {
    val theme = LocalUiTheme.current
    val context = LocalContext.current
    val billingState by store.billingStore.state.collectAsState()

    // Web parity with SubscribePage's startedRef: the checkout starts once
    // per presentation, not on every recomposition.
    var checkoutOpened by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        val checkoutUrl = components.billing.startSubscriptionCheckout(productKey) ?: return@LaunchedEffect
        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(checkoutUrl)))
        checkoutOpened = true
    }

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(theme.spacing.md),
        modifier = Modifier
            .fillMaxWidth()
            .background(theme.colors.appBg)
            .padding(24.dp),
    ) {
        val checkoutError = billingState.checkoutError
        when {
            checkoutError != null -> {
                Text(
                    text = "Checkout unavailable",
                    color = theme.colors.textPrimary,
                    fontSize = theme.typography.sizes.lg,
                    fontWeight = FontWeight.Bold,
                )
                Text(
                    text = checkoutError,
                    color = theme.colors.statusError,
                    fontSize = theme.typography.sizes.sm,
                    textAlign = TextAlign.Center,
                )
                DismissLink(label = "Back", onClick = onDismiss)
            }
            checkoutOpened -> {
                Text(
                    text = "Finish your checkout in the browser",
                    color = theme.colors.textPrimary,
                    fontSize = theme.typography.sizes.lg,
                    fontWeight = FontWeight.Bold,
                )
                Text(
                    text = "Your subscription becomes active as soon as the payment settles.",
                    color = theme.colors.textSecondary,
                    fontSize = theme.typography.sizes.sm,
                    textAlign = TextAlign.Center,
                )
                DismissLink(label = "Done", onClick = onDismiss)
            }
            else -> {
                Text(
                    text = "Starting your checkout…",
                    color = theme.colors.textPrimary,
                    fontSize = theme.typography.sizes.lg,
                    fontWeight = FontWeight.Bold,
                )
                Text(
                    text = "One moment while we prepare your subscription.",
                    color = theme.colors.textSecondary,
                    fontSize = theme.typography.sizes.sm,
                    textAlign = TextAlign.Center,
                )
                CircularProgressIndicator(color = theme.colors.accent)
            }
        }
    }
}

@Composable
private fun DismissLink(label: String, onClick: () -> Unit) {
    val theme = LocalUiTheme.current
    Text(
        text = label,
        color = theme.colors.accent,
        fontSize = theme.typography.sizes.md,
        fontWeight = FontWeight.SemiBold,
        modifier = Modifier
            .clickable(onClick = onClick)
            .padding(vertical = 4.dp, horizontal = 8.dp),
    )
}

File: ios/App/View/Billing/SubscribeView.swift
import SwiftUI

/// The subscribe flow — the native twin of web/app's SubscribePage.tsx: on
/// appear it starts a subscription checkout for the signed-in user and hands
/// them to the session's checkoutUrl (Stripe's hosted page when deployed,
/// the web app's in-app test checkout in the sandbox) in the system browser,
/// matching how the twin opens OAuth URLs. Subscription checkout is never
/// anonymous — this view only renders inside the signed-in shell (the web
/// twin routes signed-out visitors to sign-up first).
///
/// `productKey` picks a catalog plan; nil means the default plan (the web
/// twin's `?plan=<key>` query parameter). Present it from a pricing CTA or
/// the Settings Billing card's subscribe affordance.
struct SubscribeView: View {
  @EnvironmentObject private var appComponents: AppComponents
  @EnvironmentObject private var billingStore: BillingStore
  @Environment(\.uiThemeTokens) private var theme
  @Environment(\.openURL) private var openURL
  @Environment(\.dismiss) private var dismiss

  var productKey: String? = nil

  /// Web parity with SubscribePage's startedRef: the checkout starts once
  /// per presentation, not on every state change.
  @State private var hasStarted = false
  @State private var checkoutOpened = false

  var body: some View {
    VStack(spacing: 16) {
      if let error = billingStore.checkoutError {
        Text("Checkout unavailable")
          .font(.system(size: theme.typography.sizes.lg, weight: .bold))
          .foregroundStyle(theme.colors.textPrimary)
        Text(error)
          .font(.system(size: theme.typography.sizes.sm))
          .foregroundStyle(theme.colors.statusError)
          .multilineTextAlignment(.center)
        Button("Back") {
          dismiss()
        }
        .font(.system(size: theme.typography.sizes.md, weight: .semibold))
        .foregroundStyle(theme.colors.accent)
      } else if checkoutOpened {
        Text("Finish your checkout in the browser")
          .font(.system(size: theme.typography.sizes.lg, weight: .bold))
          .foregroundStyle(theme.colors.textPrimary)
        Text("Your subscription becomes active as soon as the payment settles.")
          .font(.system(size: theme.typography.sizes.sm))
          .foregroundStyle(theme.colors.textSecondary)
          .multilineTextAlignment(.center)
        Button("Done") {
          dismiss()
        }
        .font(.system(size: theme.typography.sizes.md, weight: .semibold))
        .foregroundStyle(theme.colors.accent)
      } else {
        Text("Starting your checkout…")
          .font(.system(size: theme.typography.sizes.lg, weight: .bold))
          .foregroundStyle(theme.colors.textPrimary)
        Text("One moment while we prepare your subscription.")
          .font(.system(size: theme.typography.sizes.sm))
          .foregroundStyle(theme.colors.textSecondary)
          .multilineTextAlignment(.center)
        ProgressView()
      }
    }
    .padding(24)
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(theme.colors.appBg)
    .task {
      await startCheckout()
    }
  }

  private func startCheckout() async {
    guard !hasStarted else {
      return
    }
    hasStarted = true
    guard let checkoutURL = await appComponents.billing.startSubscriptionCheckout(productKey: productKey) else {
      return
    }
    openURL(checkoutURL)
    checkoutOpened = true
  }
}
